// strict-mocha-describers/build/it-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_utils_1 = require("./test-utils");
const mocha_1 = require("mocha");
const backup_helper_1 = require("./backup-helper");
class ItHelper {
    constructor(cls, bootstrap, method) {
        this.cls = cls;
        this.bootstrap = bootstrap;
        this.method = method;
        this.createSuiteCase = (testFunction) => {
            return (description, fn) => {
                return testFunction(description, () => {
                    return fn(this.target);
                });
            };
        };
        this.beforeEach = () => {
            this.target = this.bootstrap();
            this.backupInstance(this.method);
        };
        this.afterEach = () => {
            backup_helper_1.backupHelper.restore(this.target, this.backup);
            backup_helper_1.backupHelper.restore(this.cls, this.staticBackup);
        };
    }
    createIt() {
        return test_utils_1.testUtils.setupFunction(this.createSuiteCase, mocha_1.it);
    }
    backupInstance(method) {
        this.backup = test_utils_1.testUtils.prepare(this.target, this.cls.prototype, method);
        this.staticBackup = test_utils_1.testUtils.prepare(this.cls, this.cls);
    }
}
exports.ItHelper = ItHelper;
