// strict-mocha-describers/build/static-method-describe-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const test_utils_1 = require("./test-utils");
const backup_helper_1 = require("./backup-helper");
const mocha_1 = require("mocha");
class StaticMethodDescribeHelper {
    constructor(cls) {
        this.cls = cls;
        this.createSingleStaticDescribe = (suite) => {
            return (method, fn) => {
                suite(`Static method ${method}`, () => {
                    const wrapper = {};
                    mocha_1.beforeEach(this.getBeforeEach(wrapper, method));
                    fn(mocha_1.it);
                    mocha_1.afterEach(this.getAfterEach(wrapper));
                });
            };
        };
    }
    getBeforeEach(wrapper, method) {
        return () => {
            wrapper.backup = test_utils_1.testUtils.prepare(this.cls, this.cls, method);
        };
    }
    getAfterEach(wrapper) {
        return () => {
            backup_helper_1.backupHelper.restore(this.cls, wrapper.backup);
        };
    }
    createStaticDescribe() {
        return test_utils_1.testUtils.setupFunction(this.createSingleStaticDescribe, describe);
    }
}
exports.StaticMethodDescribeHelper = StaticMethodDescribeHelper;
